### ISSUE TITLE
Performance of gradient depends on size of implicitly used array

### ISSUE THREAD

**Issue by dermesser** [FluxML/Zygote.jl#1180]
This is an odd behavior I came across while trying to learn more about RNN. I'm using the `Flux.gradient` function as entry point here, which (in my case at least) uses `Zygote` internally. Let me know if I should rather report this in the Flux repository.

This is with Zygote v0.6.36 (Flux v0.12.9) and on Julia 1.7.2.

The problem appeared when differentiating a loss function (duh). When the loss closure had access to the entire training set (~10^6 vectors of ~10 entries each), the gradient would be *super* slow, even though only a subset was ever accessed, and the gradient was only taken w.r.t. the Flux `params`. When making sure that only a `view` of the relevant training points is in scope of the closure, taking the gradient became dramatically faster (in my original case, by 500x).

I've written a small (though maybe not minimal) piece of code reproducing the behavior, modelled after a somewhat typical RNN training loop, which I intend to use for generating text (it is supposed to be quite closely reproducing the algorithm described here: https://karpathy.github.io/2015/05/21/rnn-effectiveness/). Take note of the two implementations, `loss_fast` and `loss_slow`. They do (or are supposed to do) the same calculation. `loss_slow` indexes into the whole training set. `loss_fast` is only given a view of `window = 25` data points. The second case is MUCH faster:

```
julia> run()
  # compile warm-up
  3.442409 seconds (1.59 M allocations: 168.712 MiB, 97.98% compilation time)  # loss_fast
 12.694739 seconds (103.69 M allocations: 7.131 GiB, 8.04% gc time, 28.83% compilation time)  # loss_slow
  # here's the difference
  0.077879 seconds (63.92 k allocations: 89.816 MiB)  # loss_fast
  8.436585 seconds (102.06 M allocations: 6.988 GiB, 7.58% gc time)  # loss_slow
```

My expectation is that those array fields in scope but not in use should not affect the complexity of the gradient. Upon searching around a bit, I haven't found anyone else reporting this issue before, and I haven't found it in any performance tips yet -- although it could just be that I didn't look hard enough. Therefore I ask for apologies if this is a commonly-know stupidity I've committed here (in any case, it took me an hour to figure it out).

```julia
import Flux

# stands for a reasonably complex network or function.
function network(nchars::Int; hidden::Int=128, dropout::Real=0.5)
    Flux.Chain(
               Flux.LSTM(nchars, hidden),
               Flux.Dropout(dropout),
               Flux.LSTM(hidden, hidden),
               Flux.Dropout(dropout),
               Flux.Dense(hidden, nchars),
               Flux.softmax)
end

function train(net, training_input; epochs=10, window=25, step=10, η=0.001)
    parameters = Flux.params(net)
    opt = Flux.ADAM(η)

    # Both loss functions are more or less typical implementations of RNN loss. Nevermind the nonsensical specific invocation in the last
    # line though, it doesn't matter for the observed behavior.

    # -> Fast gradient: get view of training data (25 elements = window) in gradient closure.
    loss_fast(set) = begin
        Flux.reset!(net)
        for i in 1:window
            net(set[i])
        end
        sum(Flux.binarycrossentropy(net(zeros(Float32, length(set[1]))), set[i]) for i in window+1:2window)
    end
    # -> Very slow gradient: use implicit dependency on entire training set, but only use 25 for calculations.
    loss_slow(pos) = begin
        Flux.reset!(net)
        for i in 0:window-1
            net(training_input[pos+i])
        end
        sum(Flux.binarycrossentropy(net(zeros(Float32, length(training_input[1]))), training_input[pos+i]) for i in window:2window)
    end

    inputsize = size(training_input, 1)

    # Run one epoch
    for (i, pos) in enumerate(1:step:inputsize-2window)
        # Get gradients of loss function
        set = view(training_input, pos:pos+2window)
        @time g = Flux.gradient(() -> loss_fast(set), parameters)
        @time g2 = Flux.gradient(() -> loss_slow(pos), parameters)

        # Apply gradients to parameters for gradient descent etc. ...
    end
    net
end

function run()
    net = network(35)
    data = [rand(Float32, 35, 1) for i in 1:1_000_000]

    train(net, data)
end
```

**Comment by ToucheSir** [comment]
My understanding is that Zygote is not smart enough to know whether an array being indexed should be tracked for AD. Hence it may be trying to materialize a 1 million element array (maybe even more than one) for each iteration of the backwards pass.

**Comment by dermesser** [comment]
This totally makes sense. Then the question is, how can stupid users like me be informed of this? Sometimes we don't think of it, and sometimes this can happen accidentally when such an array enters the closure unintentionally.

**Comment by mcabbott** [comment]
It's a bit of a footgun. Simpler AD systems like Tracker / ForwardDiff don't have this, as they will never track derivatives of the data array. But Zygote just differentiates the entire expression.

I believe that https://github.com/FluxML/Zygote.jl/pull/966 might help here, as the gradient of `view` would then be some lazy thunk which, since nothing consumes it, would never actually construct the big gradient array.

**Comment by dermesser** [comment]
Thank you very much for pointing me there, this is helpful! I will close this issue, as it seems that there is nothing immediately to be done.
